From 92af33324f40270f006d61215e1ecff39deb2b5e Mon Sep 17 00:00:00 2001
From: Pierre-Hugues Husson <phh@phh.me>
Date: Tue, 11 Aug 2026 09:07:54 +0000
Subject: [PATCH] Samsung tablets use SW_MACHINE_COVER rather than SW_LID

So check whether SW_MACHINE_COVER is declared, then use it
Otherwise stick to SW_LID

Should fix https://github.com/phhusson/treble_experimentations/issues/2572

[Port to A16] All anchors intact; applied verbatim.

Change-Id: I6831fc5b604aba54024cedb615952b9015fdf602
---
 .../server/input/InputManagerService.java     | 25 ++++++++++++++++---
 1 file changed, 21 insertions(+), 4 deletions(-)

diff --git a/services/core/java/com/android/server/input/InputManagerService.java b/services/core/java/com/android/server/input/InputManagerService.java
index 841cb61601ea..0d4a2b739a90 100644
--- a/services/core/java/com/android/server/input/InputManagerService.java
+++ b/services/core/java/com/android/server/input/InputManagerService.java
@@ -443,6 +443,9 @@ public class InputManagerService extends IInputManager.Stub
     /** Switch code: Microphone. When set, the mic is muted. */
     public static final int SW_MUTE_DEVICE = 0x0e;
 
+    /** Switch code: Cover. When set, the cover is closed. */
+    public static final int SW_MACHINE_COVER = 0x10;
+
     public static final int SW_LID_BIT = 1 << SW_LID;
     public static final int SW_TABLET_MODE_BIT = 1 << SW_TABLET_MODE;
     public static final int SW_KEYPAD_SLIDE_BIT = 1 << SW_KEYPAD_SLIDE;
@@ -456,6 +459,8 @@ public class InputManagerService extends IInputManager.Stub
                                            | SW_VIDEOOUT_INSERT_BIT;
     public static final int SW_CAMERA_LENS_COVER_BIT = 1 << SW_CAMERA_LENS_COVER;
     public static final int SW_MUTE_DEVICE_BIT = 1 << SW_MUTE_DEVICE;
+    public static final int SW_MACHINE_COVER_BIT = 1 << SW_MACHINE_COVER;
+    private int mSwForLid = SW_LID;
 
     // The following are layer numbers used for z-ordering the input overlay layers on the display.
     // This is used for ordering layers inside {@code DisplayContent#getInputOverlayLayer()}.
@@ -604,6 +609,16 @@ public class InputManagerService extends IInputManager.Stub
         mWiredAccessoryCallbacks = callbacks;
     }
 
+    private void checkForSwMachineCover() {
+        int machineCoverState = getSwitchState(-1 /* deviceId */, InputDevice.SOURCE_ANY, SW_MACHINE_COVER);
+        if (machineCoverState != KEY_STATE_UNKNOWN) {
+            android.util.Log.e("PHH", "Found a SW_MACHINE_COVER. Use this instead of SW_LID");
+            mSwForLid = SW_MACHINE_COVER;
+        } else {
+            android.util.Log.e("PHH", "Not found a SW_MACHINE_COVER");
+        }
+    }
+
     void registerLidSwitchCallbackInternal(@NonNull LidSwitchCallback callback) {
         synchronized (mLidSwitchLock) {
             mLidSwitchCallbacks.add(callback);
@@ -612,7 +627,8 @@ public class InputManagerService extends IInputManager.Stub
             // state will be reported as KEY_STATE_UNKNOWN. The callback will be triggered in
             // systemRunning().
             if (mSystemReady) {
-                boolean lidOpen = getSwitchState(-1 /* deviceId */, InputDevice.SOURCE_ANY, SW_LID)
+                checkForSwMachineCover();
+                boolean lidOpen = getSwitchState(-1 /* deviceId */, InputDevice.SOURCE_ANY, mSwForLid)
                         == KEY_STATE_UP;
                 callback.notifyLidSwitchChanged(0 /* whenNanos */, lidOpen);
             }
@@ -656,7 +672,8 @@ public class InputManagerService extends IInputManager.Stub
 
             // Send the initial lid switch state to any callback registered before the system was
             // ready.
-            int switchState = getSwitchState(-1 /* deviceId */, InputDevice.SOURCE_ANY, SW_LID);
+            checkForSwMachineCover();
+            int switchState = getSwitchState(-1 /* deviceId */, InputDevice.SOURCE_ANY, mSwForLid);
             for (int i = 0; i < mLidSwitchCallbacks.size(); i++) {
                 LidSwitchCallback callback = mLidSwitchCallbacks.get(i);
                 callback.notifyLidSwitchChanged(0 /* whenNanos */, switchState == KEY_STATE_UP);
@@ -2506,8 +2523,8 @@ public class InputManagerService extends IInputManager.Stub
                     + ", mask=" + Integer.toHexString(switchMask));
         }
 
-        if ((switchMask & SW_LID_BIT) != 0) {
-            final boolean lidOpen = ((switchValues & SW_LID_BIT) == 0);
+        if ((switchMask & (1 << mSwForLid)) != 0) {
+            final boolean lidOpen = ((switchValues & (1 << mSwForLid)) == 0);
             synchronized (mLidSwitchLock) {
                 if (mSystemReady) {
                     for (int i = 0; i < mLidSwitchCallbacks.size(); i++) {
-- 
2.43.0

